### DISCUSSION TITLE
How to disable jsonls in Mason? It fails to install every startup

### DISCUSSION THREAD

**Discussion by perdakovich** [LazyVim/LazyVim#526] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/89806596/229375136-d59fafcd-5c1f-4242-bac2-fdedf0d08e64.png)

I don't really need it (as I understand it requires `npm`, and I'm not okay with that)
so how do I disable it?

**Comment by dpetka2001** [comment, marked as answer]
It's configured in `nvim-lspconfig`, so maybe
```lua
return {
{
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      jsonls = {
        mason = false,
      },
    },
  },
},
}
```
could do the trick? Haven't tested it out.

**Reply by perdakovich** [reply]
![image](https://user-images.githubusercontent.com/89806596/229447549-475282a2-debc-4eb6-9938-6ce08c2e6c57.png)

Something is wrong.

btw why do I need to put commas after brackets in Lua? Where can I read more about this? It looks like nonsense to me

**Reply by dpetka2001** [reply]
You should have put in lua file along with a `return` statement. I thought if you looked at some of the examples in [lazyvim docs](https://www.lazyvim.org/) you would be able to figure it. Commas are needed because every pair of brackets is like an array/list, so it may contain more items than one. In the example above it's just one item but you should consider it good practice to get used to doing it. You should read more about lua in the official docs and maybe some neovim lua development, but I can't recommend any good recources, except from reading code in projects you like, because I'm also new to this. There is [this](https://github.com/nanotee/nvim-lua-guide) but it's deprecated in favor of the official Neovim docs, but it might give you some hints as how to proceed.

PS: Also i made a typo in my example above, it should have been `opts` instead of just opt.

**Reply by perdakovich** [reply]
thanks for explanation! so I should put commas after brackets even if there are only one item?

If you're a "new to this", it's scary to guess who I am then :-)

about the return brackets - thanks, now it all makes more sense to me. and also now it doesn't have any errors.

and for some reason your solution didn't worked, I still have these notifications on startup :thinking: 
![image](https://user-images.githubusercontent.com/89806596/229527622-1c19b17c-b757-460f-bd23-6abcbefb1abb.png)

also Aaron's solution have this line `opts = function(_, opts_)`, and your doesn't. so these are two different solutions.
how is your solution different from Aaron's one, which one is better and why?

**Reply by dpetka2001** [reply]
```lua
return {
{
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      jsonls = {
        mason = false,
      },
    },
  },
},
}
```
I did say in my previous post that I had a typo. I wrote `opt` instead of `opts`. And in your second screenshot I see that you still have it as `opt`. Anyway I tried the code in this post and does indeed does not install json-lsp. My solution is different than the other as it explicitly removes `jsonls` from auto-installing via mason. The other solution removes all the default nvim-lspconfig servers that LazyVim has configured (which also includes `lua-ls` and I'm not sure you should remove that from auto-installing since you'll be dealing with lots of Lua with Neovim/LazyVim). So I only removed `jsonls` while the other solution removed `lua-ls` as well. That's the only difference. The rest is just different syntax to achieve this.

**Reply by perdakovich** [reply]
I'm interested in "different syntax to achieve this". I am not familiar with Lua and neovim plugins either, so could you recommend me something to read to understand this "different syntax" thing? what you think I should read first: Neovim docs, neovim-lua-guide, learn lua in x minutes? or something diiferent?

sorry, forgot about "opt" thing :-(
tried with opts, everything worked, thanks! 
btw, with Aaron's solution lua lsp is still working for some reason.

also can you change your initial reply to this fixed one, with "opts" instead of "opt"? So I can mark your solution as answer

**Reply by dpetka2001** [reply]
Not sure about resources really, sorry about that. I personally started with default Lua docs (in my free time and haven't finished yet) and nanotee github repo that I linked in a previous post (but is considered deprecated in favor of official Neovim docs). Maybe do learn lua in x minutes for a fast intro and then read official Lua and Neovim docs? As for the `lua-ls` server still working, it's because it's installed by default when first starting up Lazyvim. You could uninstall it via mason manually and then check if it will re-install it in the next Lazyvim launch (which shouldn't with the other solution). Also fixed my initial post with the correct code.

**Comment by Aaron2017Li** [comment]
Hello, bro!
Try creating a lua file in the plugin dir, whose content is: 

return {
  {
    "neovim/nvim-lspconfig",
    opts = function(\_, opts\_)
      opts_.servers = {}
    end,
  },
}

As far as I know, this will disable the auto install of lua_ls and jsonls.
Hope it's helpful.

**Reply by perdakovich** [reply]
thanks, it worked.

could you explain what this line do?
`opts = function(_, opts_)`

also If I want to install some lsp server, do I need to put it here? Can't I just do it from Mason menu?

**Reply by dpetka2001** [reply]
It defines `opts` as a function that takes 2nd argument `opts_` and returns something. In this case it returns an empty opts_servers array/list which disables the default servers that get configured in nvim-lspconfig and installed via mason. You should read more about Lua if you want to understand things like that.
